### PR TITLE
[SPARK-39012][SQL]SparkSQL cast partition value does not support all data types

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Add binary type support for schema inference in SparkSQL.

### Why are the changes needed?

Spark does not support binary and boolean types when casting partition value to a target type. This PR adds the support for binary and boolean.

### Does this PR introduce _any_ user-facing change?

No. This is more like fixing a bug.

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 1]
```diff
@@ -486,6 +486,12 @@ object PartitioningUtils extends SQLConfHelper{
       DateType
     }
 
+    val binaryTry = Try {
+      val binaryValue = Hex.unhex(raw.getBytes())
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:490]: Wouldn't this infer binary type virally from all the string types?

Comment (amaliujia) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:490]: I will revert this change. Sorry I just realized this PR is not complete yet. I have converted this to a draft RP.

**Comment by cloud-fan** [comment]
LGTM, can you update the PR description?

**Comment by amaliujia** [comment]
@cloud-fan done!

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!

**Pull request closed without merging**